### PR TITLE
fix: [PL-41228]: fix adding password for an existing passwordLess smtp

### PR DESCRIPTION
**Pull request by sahibaHarness**
## Describe your changes
Earlier since in a passwordless SMTP no secret is created, it used to error out in update operation. we are checking if the secret is previously created if not we are creating on in case of secret is added. 

Tested for: 
-> Added a password in password less SMTP
-> Updation of password less SMTP
-> Updation of SMTP with password.

https://github.com/harness/harness-core/assets/112424209/fcf6f200-c34d-4db1-8974-30643ce29c8d

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55604)

### CODE REVIEW

**Comment by sahibaHarness** [comment]
trigger messagecheck

File: 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java [hunk 1]
```diff
@@ -189,8 +189,13 @@ public RestResponse<SettingAttribute> update(@PathParam("attrId") String attrId,
       secretText.setScopedToAccount(true);
       secretText.setKmsId(secretManagerConfig.getUuid());
       String secretId = existingSmtpConfig.getEncryptedPassword();
-      Boolean isSuccessful = secretManager.updateSecretText(variable.getAccountId(), secretId, secretText, true);
-      smtpConfig.setPassword(secretId.toCharArray());
+      if (secretId != null) {
+        secretManager.updateSecretText(variable.getAccountId(), secretId, secretText, true);
+        smtpConfig.setPassword(secretId.toCharArray());
+      } else {
+        log.info("Creating Secret for previously password less SMTP ");
```
Comment (pbarapatre10) [inline 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:196]: nit: lets log accountId and smtp attr id

File: 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java [hunk 2]
```diff
@@ -189,8 +189,13 @@ public RestResponse<SettingAttribute> update(@PathParam("attrId") String attrId,
...
+        smtpConfig.setPassword(secretManager.saveSecretText(variable.getAccountId(), secretText, true).toCharArray());
```
Comment (pbarapatre10) [inline 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: What happens when customer updates an existing SMTP config with password configured, and changes it to a passwordLess , shall we delete the config or it should be fine?.  To me it looks like it should be fine but please test this scenario as well, thanks

Comment (pbarapatre10) [inline 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: What happens when customer updates an existing SMTP config with password configured, and changes it to a passwordLess , shall we delete the config or it should be fine?.  To me it looks like it should be fine but please test this scenario as well, thanks

Comment (VikasMaddukuriHarness) [reply 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: This scenario i already tested a while back but please do again. This should work.

Comment (sahibaHarness) [reply 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: If we make it password less again the secrets becomes unreferenced and when we try to add a password again it fails saying  A secret exists with the proposed secret name in your account. Please choose a different name.
Should we add a deletion of the associated seceret if we are making it password less? What should the behaviour here?

Comment (VikasMaddukuriHarness) [reply 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: Yes. Please delete the secret whenever we make the smtp password less and create a new one incase we make it password type again.
@pbarapatre10 what do you think?

Comment (VikasMaddukuriHarness) [reply 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: Yes. Please delete the secret whenever we make the smtp password less and create a new one incase we make it password type again.
@pbarapatre10 what do you think?

Comment (pbarapatre10) [reply 400-rest/src/main/java/software/wings/resources/SettingResourceNg.java:197]: Yes, there are two options here @VikasMaddukuriHarness @sahibaHarness :

1. Either delete the secret when we make smtp passwordLess in Update
2. During fresh create/save of SMTP check if secret with the id exists then update it

1 is preferred as it impacts only the UPDATE flow only, Create flow won't have any impacts. 
Thanks

**Review by pbarapatre10 — APPROVED** [review]

**Pull request merged**